### PR TITLE
btrfs-assistant: init at 1.8

### PR DESCRIPTION
**Pull request by khaneliman**
## Description of changes
Adding new package btrfs-assistant to help managing btrfs filesystems.

- btrfsmaintenance doesn't exist on nixpkgs so that functionality is missing from the application. Will need to package that application to support. 
- Snapper can be disabled through override. Defaults to enabled. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (khaneliman) [inline pkgs/tools/misc/btrfs-assistant/default.nix:51]: @AndersonTorres  I'm not done with this PR, yet. But, in relation to what we were talking about in https://github.com/NixOS/nixpkgs/pull/247021 do you think we would rather explicitly change each of these files with a patch file or is this a use case that the substituteInPlace is acceptable?

Comment (AndersonTorres) [reply pkgs/tools/misc/btrfs-assistant/default.nix:52]: It looks acceptable. `substitute*` functions warn when the pattern does not find a match, quite the opposite of `sed`.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:19]:
```suggestion
}:

let
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:26]:
```suggestion
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:82]: Short descriptions 
1. should not  be self-referent
2. should not end in punctuation


```suggestion
    description = "A GUI management tool to make managing a Btrfs filesystem easier";
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:75]:
```suggestion
      --replace '/usr/bin/snapper' "${snapper}/bin/snapper"
    '';
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:61]:
```suggestion
      --replace '/usr/bin/snapper' "${snapper}/bin/snapper"
    '';
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:17]: Just to know, why not enabling it as default? I believe that it should be enabled except if there is some drawback.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:81]: Usually we "factor" a variable when it is used in more than one place. When it is used in only one place like that, we bring the `let . . . in` right near:

```suggestion
  qtWrapperArgs = let
    runtimeDeps = lib.makeBinPath ([
      coreutils
      util-linux
    ]
    ++ lib.optional enableSnapper snapper);
 in [
   "--prefix PATH : ${runtimeDeps}"
 ];
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:23]:
```suggestion
```

Slip?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (khaneliman) [reply pkgs/tools/misc/btrfs-assistant/default.nix:17]: I originally left it always enabled, but thought since it was an optional dependency in the actual program it could be optional to bundle the additional software in the derivation. Snapper is used on Btrfs filesystems to take a snapshot at regular intervals of the filesystem to allow reverting any changes to a prior point in history, which also seemed less needed in NixOS... but I suppose I was close minded in thinking of it only being used there instead of regular Linux distros using Nix.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (khaneliman) [reply pkgs/tools/misc/btrfs-assistant/default.nix:23]: Sorry, must've reintroduced during merge conflict resolution when squashing.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [reply pkgs/tools/misc/btrfs-assistant/default.nix]: There is also the issue that snapper (according to you) works on filesystem block level and for Btrfs only, whereas Nix/NixOS works at the filesystem file level.

Different tools for different purposes, despite similar.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (AndersonTorres) [inline pkgs/tools/misc/btrfs-assistant/default.nix:39]: Is this required even when `enableSnapper == false`?

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (khaneliman) [reply pkgs/tools/misc/btrfs-assistant/default.nix]: Yeah, true. I went through and tested setting up snapper because I remembered a few use cases I had for it before.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (khaneliman) [reply pkgs/tools/misc/btrfs-assistant/default.nix:39]: Well, snapper is just a utility that creates timers and a service daemon that helps manage Btrfs snapshots. Btrfs-snap I believe is the actual mechanism for Btrfs to create the snapshot.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (drupol) [inline pkgs/tools/misc/btrfs-assistant/default.nix:26]:
```suggestion
    rev = finalAttrs.version
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (drupol) [inline pkgs/tools/misc/btrfs-assistant/default.nix:1]: Can you please sort the lambda attributes ?

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (drupol) [inline pkgs/tools/misc/btrfs-assistant/default.nix:81]: Can you please sort the `meta` attributes ?

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, left me know if this is clear enough.

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (khaneliman) [reply pkgs/tools/misc/btrfs-assistant/default.nix:1]: I sorted it this way: https://github.com/NixOS/nixpkgs/pull/241997#discussion_r1282525140

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (drupol) [inline pkgs/tools/misc/btrfs-assistant/default.nix:51]:
```suggestion
      --replace '/usr/bin/snapper' ${lib.getExe snapper}
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (drupol) [inline pkgs/tools/misc/btrfs-assistant/default.nix:66]:
```suggestion
      --replace '/usr/bin/snapper' ${lib.getExe snapper}
```

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (drupol) [inline pkgs/tools/misc/btrfs-assistant/default.nix:75]:
```suggestion
      ++ lib.optionals enableSnapper [ snapper ]);
```

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/btrfs-assistant/default.nix
Comment (khaneliman) [reply pkgs/tools/misc/btrfs-assistant/default.nix:51]: Do you want me to do this same method for the bash command?

Comment (khaneliman) [reply pkgs/tools/misc/btrfs-assistant/default.nix]: https://github.com/NixOS/nixpkgs/pull/248404 created in response :P

**Review by drupol — APPROVED** [review]

**Comment by khaneliman** [comment]
Looks like checks will be blocked until snapper commit gets merged?

**Comment by drupol** [comment]
@GrahamcOfBorg build btrfs-assistant

**Comment by khaneliman** [comment]
Was just rebasing after that merge to trigger it :P

**Comment by drupol** [comment]
> Was just rebasing after that merge to trigger it :P Is that command available to anyone or just committers?

anyone

**Comment by drupol** [comment]
@GrahamcOfBorg build btrfs-assistant

**Comment by drupol** [comment]
Thanks for your contribution!

Looks like I forgot to squash... Next time I'll double check.

**Comment by khaneliman** [comment]
> Thanks for your contribution!
> 
> Looks like I forgot to squash... Next time I'll double check.

Thanks again Pol!

**Pull request merged**
